### PR TITLE
Ensure only alive leaders and followers when fetching QQ replica states

### PR DESCRIPTION
**Pull request by Ayanda-D**
## Proposed Changes

This addresses issues reported in https://github.com/rabbitmq/rabbitmq-server/discussions/12701 and https://github.com/rabbitmq/rabbitmq-server/discussions/10423 where quorum queue leader and follower FSMs have been observed to be down (`noproc`) while reporting tools such as `check_if_node_is_quorum_critical` still pass. With this patch, such terminated processes should now be reported by existing CLI tools to help avoid risks of carrying out operations such as node restarts/upgrades, when some QQ members aren't in a good state.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [ ] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc.

### CODE REVIEW

File: deps/rabbit/src/rabbit_quorum_queue.erl [hunk 1]
```diff
@@ -442,17 +442,25 @@ become_leader0(QName, Name) ->
...
-                             {K, promotable};
-                         ({K, follower, non_voter}) ->
-                             {K, non_voter};
-                         ({K, S, _}) ->
-                             %% voter or unknown
-                             {K, S};
-                         (T) ->
-                             T
+                        (T = {K, _, _}) ->
+                            case rabbit_process:is_registered_process_alive(K) of
```
Comment (kjnilsson) [inline deps/rabbit/src/rabbit_quorum_queue.erl:446]: `whereis/1` will do here, you are already on the node where the process is running.

File: deps/rabbit/src/rabbit_quorum_queue.erl [hunk 2]
```diff
@@ -442,17 +442,25 @@ become_leader0(QName, Name) ->
...
+                                true ->
+                                    to_replica_state(T);
+                                false ->
+                                    []
```
Comment (kjnilsson) [inline deps/rabbit/src/rabbit_quorum_queue.erl:450]: I'm not sure about returning nil here. is that expected?

**Review by kjnilsson — CHANGES_REQUESTED** [review]

File: deps/rabbit/src/rabbit_quorum_queue.erl (hunk 1 above)
Comment (Ayanda-D) [reply deps/rabbit/src/rabbit_quorum_queue.erl:446]: Ok, i avoided the `rabbit_process:is_process_alive/1` which does the extra distribution work. But also noticed we have `rabbit_process:is_registered_process_alive/1` which is already doing `whereis/1` with `is_pid/1` guard check on the [local node](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbit/src/rabbit_process.erl#L77-L78), which i'd still have to do on the case statement. seems to be doing same thing (if i'm seeing correct)

File: deps/rabbit/src/rabbit_quorum_queue.erl (hunk 2 above)
Comment (Ayanda-D) [reply deps/rabbit/src/rabbit_quorum_queue.erl:450]: hmm, I think the `noproc` issue, if my understanding this right, is when we have an entry in `ra_state` of a member which had already terminated, but wasnt cleaned up from ets from FSM terminate callback. Then its still listed here as if it were still active, yet not alive. But if alive=false, we should handle it as if it had been removed from `ra_state`, so returning nil here. (i'm not sure if we should also be doing a deletion from `ra_state` here when the fsm proc is not alive)

Comment (kjnilsson) [reply deps/rabbit/src/rabbit_quorum_queue.erl:450]: all other code paths return a tuple `{RegName, Status}` so returning `[] feels a bit odd, could you either return `{K, noproc}` or use lists:filtermap to remove the entry all together?

File: deps/rabbit/src/rabbit_quorum_queue.erl (hunk 1 above)
Comment (kjnilsson) [reply deps/rabbit/src/rabbit_quorum_queue.erl:446]:
```case whereis(K) of
         undefined -> ...
         Pid -> to_replica_state(T)
    end
```

I'm not sure using a helper function for the above is worth it, just causes read indirection.

Comment (Ayanda-D) [reply deps/rabbit/src/rabbit_quorum_queue.erl]: k cool, definitely easier to read. sorry for the back-&-forth on this small item (tendency has been check/use rabbit helpers 1st 😊 after finding myself rewriting some existing utils in the past 😅 ). Yep, the read indirection not necessary here 👍

File: deps/rabbit/src/rabbit_quorum_queue.erl (hunk 2 above)
Comment (Ayanda-D) [reply deps/rabbit/src/rabbit_quorum_queue.erl]: Ok,  `lists:filtermap` sounds good. this was removing/flattening out the `[]` on the conversion to a `map`, but no need for that extra step. updated 👍

**Review by kjnilsson — APPROVED** [review]
looks good, thanks @Ayanda-D

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-server/pull/12727

**Review by michaelklishin — CHANGES_REQUESTED** [review]
```shell
gmake ct-unit_quorum_queue
```

[fails repeatably](https://github.com/rabbitmq/rabbitmq-server/actions/runs/11843844517/job/33009072541?pr=12727), both on Actions and for locally for me, with

```
Node: rabbit_shard1@localhost
Case: unit_quorum_queue_SUITE:all_replica_states_includes_nonvoters
Reason: {error,
            {{badmatch,{rabbit_shard1@localhost,#{}}},
             [{unit_quorum_queue_SUITE,all_replica_states_includes_nonvoters,
                  1,
                  [{file,"unit_quorum_queue_SUITE.erl"},{line,111}]},
              {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
              {test_server,run_test_case_eval1,6,
                  [{file,"test_server.erl"},{line,1302}]},
              {test_server,run_test_case_eval,9,
                  [{file,"test_server.erl"},{line,1234}]}]}}

Node: rabbit_shard4@localhost
Case: mirrored_supervisor_SUITE:{end_per_group,broker_tests,[]}
Reason: {error,
            {{assertEqual,
                 [{module,rabbit_ct_broker_helpers},
                  {line,1218},
                  {expression,"CrashesCount"},
                  {expected,0},
                  {value,21}]},
             [{rabbit_ct_broker_helpers,find_crashes_in_logs,2,
                  [{file,"rabbit_ct_broker_helpers.erl"},{line,1218}]},
              {rabbit_ct_broker_helpers,stop_rabbitmq_nodes,1,
                  [{file,"rabbit_ct_broker_helpers.erl"},{line,1174}]},
              {rabbit_ct_helpers,run_steps,2,
                  [{file,"rabbit_ct_helpers.erl"},{line,136}]},
              {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
              {test_server,run_test_case_eval1,6,
                  [{file,"test_server.erl"},{line,1390}]},
              {test_server,run_test_case_eval,9,
                  [{file,"test_server.erl"},{line,1234}]}]}}
```

**Comment by Ayanda-D** [comment]
didnt know there were more QQ tests elsewhere :) updated, thanks for the review & QA @kjnilsson @michaelklishin  👍

**Comment by michaelklishin** [comment]
Thank you, @Ayanda-D.

**Comment by michaelklishin** [comment]
Merged in https://github.com/rabbitmq/rabbitmq-server/pull/12727.

**Pull request closed without merging**